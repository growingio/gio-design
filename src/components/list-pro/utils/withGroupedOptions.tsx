import React, { ComponentType } from 'react';
import { get, groupBy } from 'lodash';

type GetDisplayName = (x: ComponentType) => string;
export const getDisplayName: GetDisplayName = (Component) => Component.displayName || Component.name || 'Component';

const getGroupedOptions = (options: any[]) => {
  const hasNotGroup = !options.some((option: any) => {
    if (option?.groupLabel) {
      return true;
    } 
    return false;
  });

  if (hasNotGroup) {
    return options;
  }

  const groupedOptions = groupBy(options, 'groupLabel');
  return Object.keys(groupedOptions).reduce((opts: any[], groupKey: string) => {
    const group = get(groupedOptions, `${groupKey}.0.group`);
    const groupValue = get(groupedOptions, `${groupKey}.0.groupValue`);
    return opts.concat([
      {
        id: groupValue || groupKey,
        value: groupValue || groupKey,
        name: groupKey,
        label: groupKey,
        type: 'groupLabel',
        group,
      },
      ...groupedOptions[groupKey],
    ]);
  }, []);
};

export interface WithGroupedOptionsProps {
  options: any[];
  groups?: any[];
  originalOptions?: any[];
}

const withGroupedOptions = (Component: any, getGroupedFunc = getGroupedOptions): any => {
  class HOC extends React.PureComponent<any> {
    public static displayName = `GroupedOptionsProvider(${getDisplayName(Component)})`;

    public render() {
      const { options } = this.props;
      const groupedOptions = getGroupedOptions(options);
      let groups;
      if (getGroupedFunc === getGroupedOptions) {
        groups = groupedOptions;
      } else {
        groups = getGroupedFunc(options);
      }
      return (
        <Component {...this.props} options={groupedOptions} groups={groups} originalOptions={options} />
      );
    }
  }
  return HOC;
};

export default withGroupedOptions;
