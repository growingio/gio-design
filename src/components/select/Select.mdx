import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Select from './Select';

# Select 选择器

<Subtitle>
  下拉选择器，让用户做出选择。Select 与 Dropdown 的区别是 Select 承载的是选项，Dropdown 的承载的是操作。Select
  有三种尺寸的高度（大：40px、中：36px、小：30px），同 button。
</Subtitle>

## 参数说明

<ArgsTable of={Select} />

## 代码演示

### 默认

根据使用场景选用紧凑 padding 型的 Select，或者 固定宽度的 Select。Select 的下拉菜单最小宽度为 160px。

[Example](/?path=/story/functional-components-select--default)

### 多选

1. 可以通过输入及下拉选项卡进行多选。多选项以 Tag 的形式出现在输入框内。每个选项 Tag 都可以通过「X」删除。
2. Select 输入框的宽度根据页面需要设定，Select 的下拉菜单最小宽度为 160px。

[Example](/?path=/story/functional-components-select--multiple)

### 无边框背景的 Select

当页面的整洁度要求很高时，可以选用没有边框背景的 Select。

[Example](/?path=/story/functional-components-select--no-border)
