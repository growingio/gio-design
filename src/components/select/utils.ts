import React from 'react';
import { concat } from 'lodash';
import { Option, OptionProps } from './interface';

interface group {
    groupLabel?: string;
    groupValue?: string | number;
}
// ReactNode To Options
export function convertNodeToOption(node: React.ReactElement, group: group): Option {
    const {
        props: { value, children, ...restProps },
    } = node as React.ReactElement & { props: OptionProps };
    const { groupValue, groupLabel } = group;
    return { value, label: children !== undefined ? children : value, groupValue, groupLabel, ...restProps };
}

export function convertChildrenToData(nodes: React.ReactNode, group = {}): Option[] {
    let nodeOptions: Option[] = [];
    React.Children.forEach(nodes, (node: React.ReactElement) => {
        if (!React.isValidElement(node)) {
            return;
        }
        const {
            type: { isSelectOptGroup },
            props: { children, label, value },
        } = node as React.ReactElement & { type: { isSelectOptGroup?: boolean }, props: OptionProps }; // 联合类型
        if (!isSelectOptGroup) {
            // option
            nodeOptions.push(convertNodeToOption(node, group));
        } else {
            // Group
            nodeOptions = concat(nodeOptions, convertChildrenToData(children, { groupLabel: label, groupValue: value }));
        }
    });
    return nodeOptions;
}
