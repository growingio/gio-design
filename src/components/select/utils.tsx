import React from 'react';
import { concat, uniqueId } from 'lodash';
import { Option, OptionProps } from './interface';

interface group {
  groupLabel?: string;
  groupValue?: string | number;
}
const ungroupedOptionKeyPrefix = 'select_isungrouped_option_';
const ungroupedOptionKey = uniqueId(ungroupedOptionKeyPrefix);
const ungroupedOptionLabel = '未分组';
export const customOptionKeyPrefix = 'select_custom_option_';
export const customOptionKey = uniqueId(customOptionKeyPrefix);

// ReactNode To Options
export function convertNodeToOption(node: React.ReactElement, group: group): Option {
  const {
    props: { value, children, ...restProps },
  } = node as React.ReactElement & { props: OptionProps };
  const { groupValue, groupLabel } = group;
  return { value, label: children !== undefined ? children : value, groupValue, groupLabel, ...restProps };
}

export const getFlattenOptions = (data: Option[], hasGroup: boolean) => {
  const groupMap = new Map();
  if (!hasGroup) return data;
  data?.map((cur: Option) => {
    const gValue = groupMap.get(cur.groupValue);
    if (gValue) {
      const { options, ...rest } = gValue;
      return groupMap.set(cur.groupValue, {
        options: [...options, cur],
        ...rest,
      });
    }
    return groupMap.set(cur.groupValue, {
      label: cur.groupLabel,
      value: cur.groupValue,
      isSelectOptGroup: true,
      options: [cur],
    });
  });
  const flattenOption: Option[] = [];
  groupMap.forEach((value) => {
    flattenOption.push(value);
    flattenOption.push(...value.options);
  });
  return flattenOption;
};

export function convertChildrenToData(nodes: React.ReactNode, group = {}): Option[] {
  let nodeOptions: Option[] = [];
  React.Children.forEach(nodes, (node: React.ReactElement) => {
    if (!React.isValidElement(node)) {
      return;
    }
    const {
      type: { isSelectOptGroup },
      props: { children, label, value },
    } = node as React.ReactElement & { type: { isSelectOptGroup?: boolean }; props: OptionProps }; // 联合类型
    if (!isSelectOptGroup) {
      // option
      nodeOptions.push(convertNodeToOption(node, group));
    } else {
      // Group
      nodeOptions = concat(nodeOptions, convertChildrenToData(children, { groupLabel: label, groupValue: value }));
    }
  });
  return nodeOptions;
}
// optionsMap flattenOptions
export function handleOptions(
  mergedOptions: OptionProps[],
  setCacheOptions: (options: Option[]) => void,
  updateGroup: (value: boolean) => void
) {
  const isGroup = !!mergedOptions.find((v) => Object.prototype.hasOwnProperty.call(v, 'groupValue'));
  const flattenOptions: Option[] = [];
  const optionsMap = new Map();

  mergedOptions.forEach((option: Option) => {
    const { groupLabel: optionGroupLabel, groupValue: optionGroupValue, value: optionValue } = option;
    if (isGroup && !optionGroupValue && !optionGroupLabel) {
      const ungroupedOption = {
        groupLabel: ungroupedOptionLabel,
        groupValue: ungroupedOptionKey,
        ...option,
      };
      optionsMap.set(optionValue, ungroupedOption);
      flattenOptions.push(ungroupedOption);
    } else {
      optionsMap.set(optionValue, option);
      flattenOptions.push(option);
    }
  });
  updateGroup(isGroup);
  setCacheOptions(flattenOptions);
  return {
    optionsMap,
    flattenOptions,
    isGroup,
  };
}

// provide search matching hightlight;
export const defaultLabelRenderer = (input: string, prefix: string) => (
  option: Option,
  isGroup: boolean
): React.ReactNode => {
  if (isGroup || typeof option.label !== 'string') return option.title || option.label;
  const index = option.label.indexOf(input);
  return (
    <div>
      {option.label.slice(0, index)}
      <span className={`${prefix}-search-highlight`}>{option.label.slice(index, index + input.length)}</span>
      {option.label.slice(index + input.length)}
    </div>
  );
};
export const defaultOptionLabelRenderer = (value: string | number, option?: Option) =>
  option?.title || option?.label || value;
export const defaultSearchPredicate = (input: string) => (o: Option) => {
  return typeof o.label === 'string' ? o.label.includes(input) : true;
};
export const defaultMatchPredicate = (input: string) => (o: Option) => o.label === input;

export const CustomOption = (value: string | number, withGroup = false, id = customOptionKeyPrefix): Option => {
  return withGroup
    ? {
        value,
        label: value.toString(),
        groupValue: id,
        groupLabel: '自由输入',
      }
    : {
        value,
        label: value.toString(),
      };
};
