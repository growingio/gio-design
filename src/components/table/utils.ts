import { clone, has, get, join, set } from 'lodash';
import { ColumnsType, InnerColumnsType } from './interface';

export const translateInnerColumns = <RecordType>(columns: ColumnsType<RecordType>): InnerColumnsType<RecordType> =>
  clone(columns).map((cloneColumn) => {
    if (!has(cloneColumn, 'key')) {
      if (has(cloneColumn, 'dataIndex')) {
        if (Array.isArray(get(cloneColumn, 'dataIndex'))) {
          set(cloneColumn, 'key', join(get(cloneColumn, 'dataIndex'), '-'));
        } else {
          set(cloneColumn, 'key', get(cloneColumn, 'dataIndex'));
        }
      } else {
        // eslint-disable-next-line no-console
        console.warn('gio-design table: column key or dataIndex must have one');
      }
    }
    if (has(cloneColumn, 'children')) {
      set(cloneColumn, 'children', translateInnerColumns(get(cloneColumn, 'children')));
    }
    return cloneColumn;
  }) as InnerColumnsType<RecordType>;

export default {};
