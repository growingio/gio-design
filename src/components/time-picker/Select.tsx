import React, { Component } from 'react';
import classNames from 'classnames';
import raf from 'raf';
import { SelectProps } from './interface';

const scrollTo = (element: HTMLElement, to: number, duration: number) => {
  // jump to target if duration zero
  if (duration <= 0) {
    raf(() => {
      // eslint-disable-next-line no-param-reassign
      element.scrollTop = to;
    });
    return;
  }
  const difference = to - element.scrollTop;
  const perTick = (difference / duration) * 10;

  raf(() => {
    // eslint-disable-next-line no-param-reassign
    element.scrollTop += perTick;
    if (element.scrollTop === to) return;

    scrollTo(element, to, duration - 10);
  });
};

interface SelectState {
  active: boolean;
}
class Select extends Component<SelectProps, SelectState> {
  public list: any;

  public root: any;
  
  public constructor(s: SelectProps) {
    super(s);
    this.state = {
      active: false,
    };
  }

  public static state = {
    active: false,
  };

  public componentDidMount() {
    // jump to selected option
    this.scrollToSelected(0);
  }

  public componentDidUpdate(prevProps: SelectProps) {
    const { selectedIndex } = this.props;
    // smooth scroll to selected option
    if (prevProps.selectedIndex !== selectedIndex) {
      this.scrollToSelected(120);
    }
  }

  private onSelect = (value: any) => {
    const { onSelect, type } = this.props;
    onSelect(type, value);
  };

  private getOptions() {
    const { options, selectedIndex, prefixCls, onEsc } = this.props;
    return options.map((item, index) => {
      const cls = classNames({
        [`${prefixCls}-select-option-selected`]: selectedIndex === index,
        [`${prefixCls}-select-option-disabled`]: item.disabled,
      });
      const onClick = item.disabled
        ? undefined
        : () => {
            this.onSelect(item.value);
          };
      const onKeyDown = (e: any) => {
        if (e.keyCode === 13) {
          onClick && onClick();
        } else if (e.keyCode === 27) {
          onEsc && onEsc();
        }
      };

      const disprops = { disabled: item.disabled };
      return (
        <li
          // eslint-disable-next-line jsx-a11y/no-noninteractive-element-to-interactive-role
          role="button"
          onClick={onClick}
          className={cls}
          key={item.value}
          tabIndex={0}
          onKeyDown={onKeyDown}
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...disprops}
        >
          {item.value}
        </li>
      );
    });
  }

  private handleMouseEnter = (e: any) => {
    const { onMouseEnter } = this.props;
    this.setState({ active: true });
    onMouseEnter(e);
  };

  private handleMouseLeave = () => {
    this.setState({ active: false });
  };

  private saveRoot = (node: HTMLDivElement) => {
    this.root = node;
  };

  private saveList = (node: HTMLUListElement) => {
    this.list = node;
  };

  private scrollToSelected(duration: number) {
    // move to selected item
    const { selectedIndex } = this.props;
    if (!this.list) {
      return;
    }
    let index = selectedIndex;
    if (index < 0) {
      index = 0;
    }
    const topOption = this.list.children[index];
    const to = topOption.offsetTop - 8;
    scrollTo(this.root, to, duration);
  }

  public render() {
    const { prefixCls, options } = this.props;
    const { active } = this.state;

    if (options.length === 0) {
      return null;
    }
    const cls = classNames(`${prefixCls}-select`, {
      [`${prefixCls}-select-active`]: active,
    });
    return (
      <div
        className={cls}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
        ref={this.saveRoot}
      >
        <ul ref={this.saveList}>{this.getOptions()}</ul>
      </div>
    );
  }
}

export default Select;
