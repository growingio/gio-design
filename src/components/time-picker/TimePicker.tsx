import React, { Component, ReactNode } from 'react';
import Trigger from 'rc-trigger';
import moment, { Moment } from 'moment';
import classNames from 'classnames';
import Panel from './Panel';
import placements, { defaultAlign } from './placements';
import { TimePickerProps } from './interface';

function noop() {/* ... */}

function refFn(field: string, component: ReactNode) {
  this[field] = component;
}

interface TimePickerState {
  open?: boolean;
  value?: Moment;
}

class Picker extends Component<TimePickerProps, TimePickerState> {
  public static defaultProps = {
    clearText: 'clear',
    prefixCls: 'rc-time-picker',
    defaultOpen: false,
    inputReadOnly: false,
    style: {},
    className: '',
    inputClassName: '',
    popupClassName: '',
    popupStyle: {},
    align: defaultAlign,
    defaultOpenValue: moment(),
    allowEmpty: true,
    showHour: true,
    showMinute: true,
    showSecond: true,
    disabledHours: noop,
    disabledMinutes: noop,
    disabledSeconds: noop,
    hideDisabledOptions: false,
    placement: 'bottomLeft',
    onChange: noop,
    onAmPmChange: noop,
    onOpen: noop,
    onClose: noop,
    onFocus: noop,
    onBlur: noop,
    addon: noop,
    use12Hours: false,
    focusOnOpen: false,
    onKeyDown: noop,
  };

  private saveInputRef: any;

  private savePanelRef: any;

  private picker: any;

  public constructor(props: TimePickerProps) {
    super(props);
    this.saveInputRef = refFn.bind(this, 'picker');
    this.savePanelRef = refFn.bind(this, 'panelInstance');
    const { defaultOpen, defaultValue, open = defaultOpen, value = defaultValue } = props;
    this.state = {
      open,
      value,
    };
  }

  public static getDerivedStateFromProps(props: TimePickerProps, state: TimePickerState) {
    const newState: any = {};
    if ('value' in props) {
      newState.value = props.value;
    }
    if (props.open !== undefined) {
      newState.open = props.open;
    }
    return Object.keys(newState).length > 0
      ? {
          ...state,
          ...newState,
        }
      : null;
  }

  private onPanelChange = (value: Moment) => {
    this.setValue(value);
  };

  private onAmPmChange = (ampm: any) => {
    const { onAmPmChange } = this.props;
    onAmPmChange(ampm);
  };

  private onClear = (event: any) => {
    event.stopPropagation();
    this.setValue(undefined);
    this.setOpen(false);
  };

  private onVisibleChange = (open: boolean) => {
    this.setOpen(open);
  };

  private onEsc = () => {
    this.setOpen(false);
    this.focus();
  };

  private onKeyDown = (e: any) => {
    if (e.keyCode === 40) {
      this.setOpen(true);
    }
  };

  private setValue(value?: Moment) {
    const { onChange } = this.props;
    if (!('value' in this.props)) {
      this.setState({
        value,
      });
    }
    if (onChange) {
      onChange(value);
    }
  }

  private getFormat() {
    const { format, showHour, showMinute, showSecond, use12Hours } = this.props;
    if (format) {
      return format;
    }

    if (use12Hours) {
      const fmtString = [showHour ? 'h' : '', showMinute ? 'mm' : '', showSecond ? 'ss' : '']
        .filter((item) => !!item)
        .join(':');

      return fmtString.concat(' a');
    }

    return [showHour ? 'HH' : '', showMinute ? 'mm' : '', showSecond ? 'ss' : ''].filter((item) => !!item).join(':');
  }

  private getPanelElement() {
    const {
      prefixCls,
      placeholder,
      disabledHours,
      disabledMinutes,
      disabledSeconds,
      hideDisabledOptions,
      inputReadOnly,
      showHour,
      showMinute,
      showSecond,
      defaultOpenValue,
      clearText,
      addon,
      use12Hours,
      focusOnOpen,
      onKeyDown,
      hourStep,
      minuteStep,
      secondStep,
      clearIcon,
    } = this.props;
    const { value } = this.state;
    return (
      <Panel
        clearText={clearText}
        prefixCls={`${prefixCls}-panel`}
        ref={this.savePanelRef}
        value={value}
        inputReadOnly={inputReadOnly}
        onChange={this.onPanelChange}
        onAmPmChange={this.onAmPmChange}
        defaultOpenValue={defaultOpenValue}
        showHour={showHour}
        showMinute={showMinute}
        showSecond={showSecond}
        onEsc={this.onEsc}
        format={this.getFormat()}
        placeholder={placeholder}
        disabledHours={disabledHours}
        disabledMinutes={disabledMinutes}
        disabledSeconds={disabledSeconds}
        hideDisabledOptions={hideDisabledOptions}
        use12Hours={use12Hours}
        hourStep={hourStep}
        minuteStep={minuteStep}
        secondStep={secondStep}
        addon={addon}
        focusOnOpen={focusOnOpen}
        onKeyDown={onKeyDown}
        clearIcon={clearIcon}
      />
    );
  }

  private getPopupClassName() {
    const { showHour, showMinute, showSecond, use12Hours, prefixCls, popupClassName } = this.props;
    let selectColumnCount = 0;
    if (showHour) {
      selectColumnCount += 1;
    }
    if (showMinute) {
      selectColumnCount += 1;
    }
    if (showSecond) {
      selectColumnCount += 1;
    }
    if (use12Hours) {
      selectColumnCount += 1;
    }
    // Keep it for old compatibility
    return classNames(
      popupClassName,
      {
        [`${prefixCls}-panel-narrow`]: (!showHour || !showMinute || !showSecond) && !use12Hours,
      },
      `${prefixCls}-panel-column-${selectColumnCount}`
    );
  }

  private setOpen(open: boolean) {
    const { onOpen, onClose } = this.props;
    const { open: currentOpen } = this.state;
    if (currentOpen !== open) {
      if (!('open' in this.props)) {
        this.setState({ open });
      }
      if (open) {
        onOpen && onOpen({ open });
      } else {
        onClose && onClose({ open });
      }
    }
  }

  private focus() {
    this.picker.focus();
  }

  // private blur() {
  //   this.picker.blur();
  // }

  private renderClearButton() {
    const { value } = this.state;
    const { prefixCls, allowEmpty, clearIcon, clearText, disabled } = this.props;
    if (!allowEmpty || !value || disabled) {
      return null;
    }

    if (React.isValidElement(clearIcon)) {
      const { onClick } = clearIcon.props || {};
      const clearIconNode = React.cloneElement(clearIcon, {
        onClick: (...args: any) => {
          if (onClick) {
            onClick(...args);
          }
          if (args && args[0]) {
            this.onClear(args[0]);
          }
        },
      });

      return (
        <span role="button" className={`${prefixCls}-clear`}>
          {clearIconNode}
        </span>
      );
    }

    return (
      <button type="button" className={`${prefixCls}-clear`} title={clearText} onClick={this.onClear} tabIndex={0} aria-hidden="true">
        {clearIcon}
      </button>
    );
  }

  public render() {
    const {
      prefixCls,
      placeholder,
      placement = 'bottomRight',
      id,
      disabled,
      transitionName,
      className,
      inputReadOnly,
      name,
      inputIcon,
      inputClassName,
      getPopupContainer,
      autoComplete,
      onFocus,
      onBlur,
      autoFocus,
      style,
      align,
      popupStyle,
    } = this.props;
    const { open, value } = this.state;
    const popupClassName = this.getPopupClassName();
    return (
      <Trigger
        prefixCls={`${prefixCls}-panel`}
        popupClassName={popupClassName}
        popupStyle={popupStyle}
        popup={this.getPanelElement()}
        popupAlign={align}
        builtinPlacements={placements}
        popupPlacement={placement}
        action={disabled ? [] : ['click']}
        destroyPopupOnHide
        getPopupContainer={getPopupContainer}
        popupTransitionName={transitionName}
        popupVisible={open}
        onPopupVisibleChange={this.onVisibleChange}
      >
        <span className={classNames(prefixCls, className)} style={style}>
          <input
            className={classNames(`${prefixCls}-input`, inputClassName, {
              [`${prefixCls}-input-focus`]: open,
            })}
            ref={this.saveInputRef}
            type="text"
            placeholder={placeholder}
            name={name}
            onKeyDown={this.onKeyDown}
            disabled={disabled}
            value={(value && value.format(this.getFormat())) || ''}
            autoComplete={autoComplete}
            onFocus={onFocus}
            onBlur={onBlur}
            autoFocus={autoFocus} // eslint-disable-line jsx-a11y/no-autofocus
            onChange={noop}
            readOnly={!!inputReadOnly}
            id={id}
            required
          />
          <span role="button" className={`${prefixCls}-addon-icon`}>
            {inputIcon}
          </span>
          {this.renderClearButton()}
        </span>
      </Trigger>
    );
  }
}

export default Picker;
