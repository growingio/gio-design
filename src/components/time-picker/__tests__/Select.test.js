import React from 'react';
import { mount } from 'enzyme';
import ReactDOM from 'react-dom';
import moment from 'moment';
import { noop } from 'lodash';
import { clickInput, clickSelectItem, matchValue, matchAll, escapeSelected } from './util';
import Picker from '../index';
import Select from '../Select';
import TimePicker from '../TimePicker';
import { sleep } from '../../../utils/test';

// 打印快照
describe('Testing select', () => {
  it('should render a DOM', () => {
    const wrapper = mount(<Picker className="test-cls" />);
    expect(wrapper.find('.gio-time-picker')).toHaveLength(1);
  });

  it('should render null', () => {
    expect(
      mount(
        <Select
          prefixCls="gio-time-picker-select"
          options={[]}
          type="ampm"
          onSelect={noop}
          onMouseEnter={noop}
          onEsc={noop}
        />
      )
    ).toMatchSnapshot();
  });

  it('disable', () => {
    expect(
      mount(
        <Select
          prefixCls="gio-time-picker-select"
          options={[{ disabled: true, value: 'test1' }, { value: 'test2' }]}
          selectedIndex={1}
          type="ampm"
          onSelect={noop}
          onMouseEnter={noop}
          onEsc={noop}
        />
      )
    ).toMatchSnapshot();
  });
  it('should render a DOM', () => {
    const wrapper = mount(<Picker className="test-cls" />);
    expect(wrapper.find('.gio-time-picker')).toHaveLength(1);
  });

  it('should call onSelect function whit enter key', () => {
    const onSelect = jest.fn();
    const wrapper = mount(
      <Select
        prefixCls="gio-time-picker-select"
        options={[{ value: 'test1' }, { value: 'test2' }]}
        selectedIndex={1}
        type="ampm"
        onSelect={onSelect}
        onMouseEnter={() => {}}
        onEsc={() => {}}
      />
    );
    wrapper.find('.gio-time-picker-select-select li').at(0).simulate('keydown', { keyCode: 13 });
    expect(onSelect).toHaveBeenCalled();
  });
});

describe('Select', () => {
  let container;

  function renderPicker(props) {
    const showSecond = true;
    const format = 'HH:mm:ss';

    return mount(
      <TimePicker format={format} showSecond={showSecond} defaultValue={moment('01:02:04', format)} {...props} />
    );
  }

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(container);
    document.body.removeChild(container);
  });

  describe('select panel', () => {
    it('select panel reacts to mouseenter and mouseleave correctly', async () => {
      const picker = renderPicker({ use12Hours: true });
      clickInput(picker);

      const re = /(^|\s+)gio-time-picker-panel-select-active(\s+|$)/;

      expect(re.test(picker.find('.gio-time-picker-panel-select').at(0).instance().className)).toBeFalsy();

      picker.find('.gio-time-picker-panel-select').at(0).simulate('mouseEnter');
      expect(re.test(picker.find('.gio-time-picker-panel-select').at(0).instance().className)).toBeTruthy();

      picker.find('.gio-time-picker-panel-select').at(0).simulate('mouseLeave');
      expect(re.test(picker.find('.gio-time-picker-panel-select').at(0).instance().className)).toBeFalsy();

      picker.find('.gio-time-picker-panel-select').at(1).simulate('mouseEnter');

      picker.find('.gio-time-picker-panel-select').at(2).simulate('mouseEnter');

      picker.find('.gio-time-picker-panel-select').at(3).simulate('mouseEnter');
    });

    it('shows only numbers according to step props', async () => {
      const picker = renderPicker({
        hourStep: 5,
        minuteStep: 15,
        secondStep: 21,
      });
      clickInput(picker);

      const selectors = picker.find('.gio-time-picker-panel-select');

      const hourSelector = selectors.at(0);
      const minuteSelector = selectors.at(1);
      const secondSelector = selectors.at(2);

      const hours = hourSelector.find('li').map((node) => node.text());
      expect(hours).toEqual(['00', '05', '10', '15', '20']);

      const minutes = minuteSelector.find('li').map((node) => node.text());
      expect(minutes).toEqual(['00', '15', '30', '45']);

      const seconds = secondSelector.find('li').map((node) => node.text());
      expect(seconds).toEqual(['00', '21', '42']);
    });
  });

  describe('select number', () => {
    it('select number correctly', async () => {
      const picker = renderPicker();
      expect(picker.state().open).toBeFalsy();

      clickInput(picker);
      expect(picker.state().open).toBeTruthy();

      expect(picker.find('.gio-time-picker-panel-select').length).toBe(3);
    });
  });

  describe('select in 12 hours mode', () => {
    it('renders correctly', async () => {
      const picker = renderPicker({
        use12Hours: true,
        defaultValue: moment().hour(14).minute(0).second(0),
        showSecond: false,
        format: undefined,
      });

      expect(picker.state().open).toBeFalsy();
      clickInput(picker);
      expect(picker.state().open).toBeTruthy();

      matchValue(picker, '2:00 pm');

      expect(picker.find('.gio-time-picker-panel-select').length).toBe(3);
    });

    it('renders 12am correctly', async () => {
      const picker = renderPicker({
        use12Hours: true,
        defaultValue: moment().hour(0).minute(0).second(0),
        showSecond: false,
        format: undefined,
      });
      expect(picker.state().open).toBeFalsy();
      clickInput(picker);
      expect(picker.state().open).toBeTruthy();

      expect(picker.find('.gio-time-picker-panel-select').length).toBe(3);
    });

    it('renders 5am correctly', async () => {
      const picker = renderPicker({
        use12Hours: true,
        defaultValue: moment().hour(0).minute(0).second(0),
        showSecond: false,
        format: undefined,
      });
      expect(picker.state().open).toBeFalsy();
      clickInput(picker);
      expect(picker.state().open).toBeTruthy();

      matchValue(picker, '12:00 am');
      clickSelectItem(picker, 0, 3);

      matchValue(picker, '3:00 am');
    });

    it('renders 12am/pm correctly', async () => {
      const picker = renderPicker({
        use12Hours: true,
        defaultValue: moment().hour(0).minute(0).second(0),
        showSecond: false,
        format: undefined,
      });

      expect(picker.state().open).toBeFalsy();
      clickInput(picker);
      expect(picker.state().open).toBeTruthy();

      matchValue(picker, '12:00 am');

      clickSelectItem(picker, 2, 1);
      matchValue(picker, '12:00 pm');

      clickSelectItem(picker, 2, 0);
      matchValue(picker, '12:00 am');
    });

    it('renders uppercase AM correctly', async () => {
      const picker = renderPicker({
        use12Hours: true,
        defaultValue: moment().hour(0).minute(0).second(0),
        showSecond: false,
        format: 'h:mm A',
      });

      expect(picker.state().open).toBeFalsy();
      clickInput(picker);
      expect(picker.state().open).toBeTruthy();

      matchValue(picker, '12:00 AM');

      clickSelectItem(picker, 2, 1);
      matchValue(picker, '12:00 PM');

      clickSelectItem(picker, 2, 0);
      matchValue(picker, '12:00 AM');
    });
  });

  it('escape closes popup', async () => {
    const picker = renderPicker();

    expect(picker.state().open).toBeFalsy();
    clickInput(picker);
    expect(picker.state().open).toBeTruthy();

    clickSelectItem(picker, 1, 1);
    escapeSelected(picker);

    expect(picker.state().open).toBeFalsy();
  });
});

it('onScroll', async () => {
  const options = new Array(100).fill(0).map((_, index) => ({ value: `test${index}`, label: `test${index}` }));
  const wrapper = mount(
    <Select
      prefixCls="gio-time-picker-select"
      options={options}
      selectedIndex={88}
      type="ampm"
      onSelect={() => {}}
      onMouseEnter={() => {}}
      onEsc={() => {}}
    />
  );
  await sleep(50);
  wrapper.instance().scrollToSelected(100);
  await sleep(50);
  expect(wrapper.render).toMatchSnapshot();
});
