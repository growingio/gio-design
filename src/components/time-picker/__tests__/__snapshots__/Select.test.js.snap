// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing select disable 1`] = `ReactWrapper {}`;

exports[`Testing select should render null 1`] = `ReactWrapper {}`;
